Add unit tests for SVG path parser and fix bugs in parser

YenPeiChen07 opened the pull request:
This PR adds unit tests for SVG path parse to test parsing logic and also fix bugs in parser, which is related to #504 and #505.

In SVG, it defines[ 6 types of path commands](https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Attribute/d#path_commands) in `d` attribute. I added unit test based on these 6 types and test multiple closed paths in a `d` attribute.

The parser works fine only for a SVG consisting of `<path>` elements. For some with [basic shapes](https://developer.mozilla.org/en-US/docs/Web/SVG/Tutorials/SVG_from_scratch/Basic_shapes), it is not currently supported.

- SVG consisting of `<path>` elements: [Apple](https://www.svgrepo.com/svg/484901/apple) 
<img width="350" alt="Screenshot 2025-04-23 at 10 54 57 AM" src="https://github.com/user-attachments/assets/b2092910-5cfa-4b9b-bc21-afac2049859a" />
<img width="250" alt="Screenshot 2025-04-23 at 10 56 02 AM" src="https://github.com/user-attachments/assets/fc8fe8c3-aac8-408a-9ddb-20b9f5ddb028" />


 - SVG consisting of `<path>` elements and basic shapes `<circle>`: [Chef Man Cap](https://www.svgrepo.com/svg/533810/chef-man-cap)
<img width="400" alt="Screenshot 2025-04-23 at 11 05 09 AM" src="https://github.com/user-attachments/assets/55884a08-6c26-4952-990a-3b93a0eee61d" />
<img width="250" alt="Screenshot 2025-04-23 at 11 05 42 AM" src="https://github.com/user-attachments/assets/16c066b9-ceba-4cf5-a7dd-46e32e74162e" />

YenPeiChen07 committed:
add unit tests for svg path parser and fix bugs

yungyuc commented on modmesh/pilot/_svg.py:
```diff
@@ -148,191 +148,272 @@ def calc_vertices(self):
-                #   m dx dy (relative position to current position)
-                x, y = coords[i], coords[i+1]
+                # Move position:
+                #   command: [M|m] (relative position in lowercase command)
```
Clear and comment.

YenPeiChen07 commented on modmesh/pilot/_svg.py:
```diff
@@ -434,26 +517,37 @@ def _load_svg_file(self, filename):
-                world.add_segment(s)
+            for j in range(len(sp2d[i])):
+                # the points are reflected to x-axis: (x, y) -> (x, -y)
+                world.add_segment(Point(sp2d[i].x0_at(j), -sp2d[i].y0_at(j)),
```
Reverse y-axis values when adding segments to align with current rendering coordinate system.

yungyuc commented on modmesh/pilot/_svg.py:
```diff
@@ -148,191 +148,272 @@ def calc_vertices(self):
                 current_pos = start_pos
                 i += 1
             else:
+                # [TODO] raise a value error
```
Why not raise the `ValueError` with this PR?

YenPeiChen07 commented on modmesh/pilot/_svg.py:
```diff
@@ -148,191 +148,272 @@ def calc_vertices(self):
+                    end = Point(x_end, y_end, 0)
+                    sp2d.append(Segment(current_pos, end))
+
+                    current_pos = end
```
Set the end point as current position after adding a segment, which fixes incorrect rendering of vertical lines.

yungyuc commented on modmesh/pilot/_svg.py:
```diff
@@ -376,6 +456,9 @@ def parse(self):
             d_attr = elmnt.attrib.get('d', '')
             fill_attr = elmnt.attrib.get('fill', '')
             paths.append(EPath(d_attr=d_attr, fill_attr=fill_attr))
+        """ d_attr = pathElements[0].attrib.get('d', '')
```
Is it dead code?  If so, delete it.  If not, make it work or add proper comments.

YenPeiChen07 commented on modmesh/pilot/_svg.py:
```diff
@@ -148,191 +148,272 @@ def calc_vertices(self):
+                # Draw lines from current position
+                #   command: [L|l] (relative position in lowercase command)
+                #   parameter: (dx1, dy1)+
+                while i + 1 < len(coords):
```
Support subsequent coordinate pair(s) for implicit LineTo (`L`/`l`) command(s).

yungyuc reviewed (approved):
LGTM.  But CI needs to be cleaned.

YenPeiChen07 commented on modmesh/pilot/_svg.py:
```diff
@@ -148,191 +148,272 @@ def calc_vertices(self):
                 i += 2
+
+                # implicit line-to command
+                while i + 1 < len(coords):
```
As described in [moveTo command](https://developer.mozilla.org/en-US/docs/Web/SVG/Reference/Attribute/d#moveto_path_commands):

> Any subsequent coordinate pair(s) are interpreted as parameter(s) for implicit absolute LineTo (L) command(s) 

It is required to add segments in `M` and `m` commands.

yungyuc replied:
Is it covered by a unit test?

YenPeiChen07 commented on modmesh/pilot/_svg.py:
```diff
@@ -148,191 +148,272 @@ def calc_vertices(self):
-                # draws a horizontal line
-                # for 'H', coord[0] is absolute position
-                # for 'h', coord[0] is relative position from current position
+            elif cmd in ('Q', 'q'):
```
Support quadratic Bézier curve commands (`Q`, `q`, `T`, and `t`).

yungyuc committed:
Move non-Qt (GUI) code to `modmesh.plot`

`modmesh.plot` is a new sub-package to hold code for non-GUI drawing Python code.

yungyuc commented on modmesh/plot/__init__.py:
@YenPeiChen07 This is my attempt to separate the non-GUI part here.

yungyuc reviewed (commented):
I pushed a commit to separate the GUI and non-GUI part.
